Content.Server/_Moffstation/GameTicking/Rules/GunGameRuleSystem.cs

```diff
@@ -0,0 +1,229 @@
+public sealed class GunGameRuleSystem : GameRuleSystem<GunGameRuleComponent>
+{
+    [Dependency] private readonly IPlayerManager _player = default!;
+    [Dependency] private readonly MindSystem _mind = default!;
+    [Dependency] private readonly OutfitSystem _outfitSystem = default!;
+    [Dependency] private readonly RespawnRuleSystem _respawn = default!;
+    [Dependency] private readonly RoundEndSystem _roundEnd = default!;
+    [Dependency] private readonly StationSpawningSystem _stationSpawning = default!;
+    [Dependency] private readonly IPrototypeManager _proto = default!;
+    [Dependency] private readonly SharedHandsSystem _hands = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<PlayerBeforeSpawnEvent>(OnBeforeSpawn);
+        SubscribeLocalEvent<PlayerSpawnCompleteEvent>(OnSpawnComplete);
+        SubscribeLocalEvent<KillReportedEvent>(OnKillReported);
+    }
+
+    private void OnBeforeSpawn(PlayerBeforeSpawnEvent ev)
+    {
+        var query = EntityQueryEnumerator<GunGameRuleComponent, RespawnTrackerComponent, GameRuleComponent>();
+        while (query.MoveNext(out var uid, out var gunGame, out var tracker, out var rule))
+        {
+            if (!GameTicker.IsGameRuleActive(uid, rule))
+                continue;
+
+            var newMind = _mind.CreateMind(ev.Player.UserId, ev.Profile.Name);
+            _mind.SetUserId(newMind, ev.Player.UserId);
+
+            var mobMaybe = _stationSpawning.SpawnPlayerCharacterOnStation(ev.Station, null, ev.Profile);
+            DebugTools.AssertNotNull(mobMaybe);
+            var mob = mobMaybe!.Value;
+
+            _mind.TransferTo(newMind, mob);
+            _outfitSystem.SetOutfit(mob, gunGame.Gear);
+            EnsureComp<KillTrackerComponent>(mob);
+            EnsureComp<GunGameRewardTrackerComponent>(mob);
+            SpawnCurrentWeapon(ev.Player.UserId, gunGame);
+            _respawn.AddToTracker(ev.Player.UserId, (uid, tracker));
+
+            ev.Handled = true;
+            break;
+        }
+    }
+
+    private void OnSpawnComplete(PlayerSpawnCompleteEvent ev)
+    {
+        EnsureComp<KillTrackerComponent>(ev.Mob);
+        EnsureComp<GunGameRewardTrackerComponent>(ev.Mob);
+        var query = EntityQueryEnumerator<GunGameRuleComponent, RespawnTrackerComponent, GameRuleComponent>();
+        while (query.MoveNext(out var uid, out _, out var tracker, out var rule))
+        {
+            if (!GameTicker.IsGameRuleActive(uid, rule))
+                continue;
+            _respawn.AddToTracker((ev.Mob, null), (uid, tracker));
+        }
+    }
+
+    /// <summary>
+    /// Activates when a kill is reported, ev.Entity corresponds to the person who was killed.
+    /// </summary>
+    private void OnKillReported(ref KillReportedEvent ev)
+    {
+        // Don't want other players picking up somebody's gun
+        DeleteCurrentWeapons(EnsureComp<GunGameRewardTrackerComponent>(ev.Entity));
```

Comment:
Seems kinda weird to `EnsureComp` to iterate over its definitely empty contents, but eh.